lib/nethealth/nethealth.go

```diff
@@ -0,0 +1,590 @@
+		promPeerRequest: promPeerRequest,
+		selector:        labelSelector,
+		triggerResync:   make(chan bool, 1),
+		rxMessage:       make(chan messageWrapper, 100),
+		peers:           make(map[string]*peer),
+		addrToPeer:      make(map[string]string),
+	}, nil
+}
+
+// Server is an instance of nethealth that is running on each node responsible for sending and responding to heartbeats.
+type Server struct {
+	logrus.FieldLogger
+
+	config     Config
+	clock      clockwork.Clock
+	conn       *icmp.PacketConn
+	httpServer *http.Server
+	selector   labels.Selector
+
+	// rxMessage is a processing queue of received echo responses
+	rxMessage     chan messageWrapper
+	triggerResync chan bool
+
+	peers      map[string]*peer
+	addrToPeer map[string]string
+
+	client kubernetes.Interface
+
+	promPeerRTT     *prometheus.HistogramVec
+	promPeerTimeout *prometheus.CounterVec
+	promPeerRequest *prometheus.CounterVec
+}
+
+type peer struct {
+	name        string
+	addr        net.Addr
+	echoCounter int
+	echoTime    time.Time
+	echoTimeout bool
+
+	status           string
+	lastStatusChange time.Time
+}
+
+type messageWrapper struct {
+	message  *icmp.Message
+	rxTime   time.Time
+	peerAddr net.Addr
+}
+
+// Start sets up the server and begins normal operation
+func (s *Server) Start() error {
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	s.client, err = kubernetes.NewForConfig(config)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	s.conn, err = icmp.ListenPacket("ip4:icmp", "0.0.0.0")
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	s.clock = clockwork.NewRealClock()
+	go s.loop()
+	go s.loopServiceDiscovery()
+	go s.serve()
+
+	mux := http.ServeMux{}
+	mux.Handle("/metrics", promhttp.Handler())
+	s.httpServer = &http.Server{Addr: fmt.Sprint(":", s.config.PrometheusPort), Handler: &mux}
+	go func() {
+		if err := s.httpServer.ListenAndServe(); err != http.ErrServerClosed {
+			s.Fatalf("ListenAndServe(): %s", err)
+		}
+	}()
+
+	s.Info("Started nethealth with config:")
+	s.Info("  PrometheusPort: ", s.config.PrometheusPort)
+	s.Info("  Namespace: ", s.config.Namespace)
+	s.Info("  NodeName: ", s.config.NodeName)
+	s.Info("  Selector: ", s.selector)
+	s.Info("  ServiceDiscoveryQuery: ", s.config.ServiceDiscoveryQuery)
+
+	return nil
+}
+
+// loop is the main processing loop for sending/receiving heartbeats.
+func (s *Server) loop() {
+	heartbeatTicker := s.clock.NewTicker(heartbeatInterval)
+	defer heartbeatTicker.Stop()
+
+	resyncTicker := s.clock.NewTicker(resyncInterval)
+	defer resyncTicker.Stop()
+
+	for {
+		select {
+		//
+		// Re-sync cluster peers
+		//
+		case <-resyncTicker.Chan():
+			err := s.resyncPeerList()
+			if err != nil {
+				s.WithError(err).Error("Unexpected error re-syncing the list of peer nodes.")
+			}
+
+			err = s.resyncNethealthPods()
+			if err != nil {
+				s.WithError(err).Error("Unexpected error re-syncing the list of peer pods.")
+			}
+		case <-s.triggerResync:
```

Comment:
So the resync interval is 15 minutes - what happens when a node is removed from the cluster until the next resync? The other nodes will keep trying pinging it and failing? Will it trigger any alerts?

Reply:
It depends on how it goes down. If the pod is deleted, that should trigger the service discovery change that will prevent sending any pings. However, as long as the node object exists, it will still count as down for pings. If the node object is deleted at the same time as the pod in kubernetes, it means the node will be cleaned up at that time as well within the nethealth.

However, if the pod get's deleted before the node, or it is otherwise missed somehow, then the latest this will be discovered is after 15 minutes.

This works in a similar way for when a node is added. When a node is added it doesn't take 15 minutes, the pings start as soon as coredns indicates that a new nethealth pod is available in the cluster on it's 15 second timer.